sdk/couchbase-core/src/memdx/response.rs

```diff
@@ -403,6 +403,861 @@ impl TryFromClientResponse for GetResponse {
+        let packet = resp.packet();
+        let status = packet.status;
+
+        if status == Status::KeyExists {
+            return Err(ServerError::new(
+                ServerErrorKind::CasMismatch,
+                resp.packet(),
+                resp.local_addr(),
+                resp.peer_addr(),
+            )
+                .into());
+        } else if status == Status::TooBig {
+            return Err(ServerError::new(
+                ServerErrorKind::TooBig,
+                resp.packet(),
+                resp.local_addr(),
+                resp.peer_addr(),
+            )
+                .into());
+        } else if status == Status::NotStored {
+            return Err(ServerError::new(
+                    ServerErrorKind::KeyNotFound,
+                    resp.packet(),
+                    resp.local_addr(),
+                    resp.peer_addr(),
+                )
+                    .into());
+        } else if status == Status::Locked {
+            return Err(ServerError::new(
+                ServerErrorKind::Locked,
+                resp.packet(),
+                resp.local_addr(),
+                resp.peer_addr(),
+            )
+                .into());
+        } else if status != Status::Success {
+            return Err(OpsCrud::decode_common_error(
+                resp.packet(),
+                resp.local_addr(),
+                resp.peer_addr(),
+            ));
+        }
+
+        let mutation_token = if let Some(extras) = &packet.extras {
+            if extras.len() != 16 {
+                return Err(ErrorKind::Protocol {
+                    msg: "Bad extras length".to_string(),
+                }
+                    .into());
+            }
+            let mut extras = Cursor::new(extras);
+
+            Some(MutationToken {
+                vbuuid: extras
+                    .read_u64::<BigEndian>()
+                    .map_err(|e| Error::from(ErrorKind::Protocol { msg: e.to_string() }))?,
+                seqno: extras
+                    .read_u64::<BigEndian>()
+                    .map_err(|e| Error::from(ErrorKind::Protocol { msg: e.to_string() }))?,
+            })
+        } else {
+            None
+        };
+
+        let server_duration = if let Some(f) = &packet.framing_extras {
+            decode_res_ext_frames(f)?
+        } else {
+            None
+        };
+
+        Ok(AppendResponse {
+            cas: packet.cas.unwrap_or_default(),
+            mutation_token,
+            server_duration,
+        })
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct PrependResponse {
+    pub cas: u64,
+    pub mutation_token: Option<MutationToken>,
+    pub server_duration: Option<Duration>,
+}
+
+impl TryFromClientResponse for PrependResponse {
+    fn try_from(resp: ClientResponse) -> Result<Self, Error> {
+        let packet = resp.packet();
+        let status = packet.status;
+
+        if status == Status::KeyExists {
```

Comment:
Gocbcorex also checks that `request.cas > 0` before returning CasMismatch, but we don't have too easy a way to do that here.. is it necessary?

Reply:
I'm not 100% sure why gocbcorex has that check. This is another reason why I want to refactor all of this to be callback based so that we have scope on the request at this point. For now I'd say go with this and file an issue to add the cas check later.

Reply:
👍 Created RSCBC-46